Add afmoe model

alyosha-swamy opened the pull request:
## Summary

This PR adds support for the **AFMoE (Arcee Foundational Mixture of Experts)** model architecture for the upcoming **Trinity-Mini** and **Trinity-Nano** releases. AFMoE is a decoder-only transformer model featuring a sparse Mixture of Experts (MoE) approach, combining token-choice routing with shared experts and several architectural innovations for efficient inference and improved performance.

## Model Description

AFMoE features the following key architectural components:

- **Mixture of Experts with Shared Experts**: Combines routed experts (activated per-token via learned routing) with always-active shared experts for stable base computation

- **Token-Choice Routing**: Uses sigmoid or softmax-based routing with normalization and scaling for expert selection

- **Q/K Normalization and Gating**: Applies RMSNorm to query and key projections and uses sigmoid gating on attention outputs for improved training stability

- **Hybrid Attention Patterns**: Alternates between sliding window attention and full attention across layers for efficiency with long contexts

- **Dual Normalization**: Uses pre- and post-normalization around both attention and MLP blocks for training stability

- **Configurable Dense Layers**: Allows initial layers to use dense MLPs before transitioning to sparse MoE layers (`num_dense_layers`)

## Implementation Details

- **Modular implementation** leveraging transformers' modular architecture:

  - Efficient `AfmoeRMSNorm` for layer normalization

  - `AfmoeRotaryEmbedding` for positional encoding

  - `AfmoeAttention` class implementing Q/K normalization and output gating

  - `AfmoeTokenChoiceRouter` for expert selection

  - `AfmoeMoE` class implementing shared + routed experts architecture

  - `AfmoeDecoderLayer` integrating attention and MoE blocks with dual normalization

## Testing

- Added comprehensive test suite following standard transformers test patterns
- Tests for core functionality:
  - Model initialization and weight loading
  - Forward and backward passes
  - Attention mechanism (sliding window + full attention patterns)
  - MoE routing and expert selection
  - RoPE embeddings
  - KV cache compatibility
- Integration tests with example checkpoints
- Verified compatibility with existing transformer infrastructure
- Model loading and inference verified with `arcee-ai/Trinity-Mini`

## Documentation

- Comprehensive model documentation in `docs/source/en/model_doc/afmoe.md`
- Detailed architecture descriptions and usage examples
- All configuration parameters documented with clear descriptions
- Example code for both Pipeline and AutoModel usage patterns

alyosha-swamy committed:
Add AFMoE model support

alyosha-swamy committed:
Merge remote-tracking branch 'upstream/main' into add_afmoe_model

ArthurZucker commented on src/transformers/models/afmoe/modular_afmoe.py:
```diff
@@ -0,0 +1,640 @@
+        if self.score_func == "sigmoid":
+            scores = torch.sigmoid(scores.to(torch.float32))
+        else:
+            scores = F.softmax(scores.to(torch.float32), dim=-1)
```
in general we get rid of stuff that is not part of the released model, so if the released model uses only one of them, let's remove the other code pathes

ArthurZucker commented on src/transformers/models/afmoe/modular_afmoe.py:
```diff
@@ -0,0 +1,640 @@
+        top_scores, selected_experts = self.router(hidden_states, self.expert_bias)
+
+        # Process through shared experts
+        if self.shared_experts is not None:
```
same comment, is this used by the released model or not?

ArthurZucker replied:
not adressed

ArthurZucker replied:
In that case the arch should be different! Use normal MLP for mlp and expert for expert! 🤗 
You can set the first layer then just do += we want to avoid codepathes as much as possible

ArthurZucker commented on src/transformers/models/afmoe/modular_afmoe.py:
```diff
@@ -0,0 +1,640 @@
+
+        # Apply muP input scaling if enabled
+        if self.config.mup_enabled:
+            hidden_states = hidden_states * (self.config.hidden_size**0.5)
```
the scaling should be done on the weight side if its constant (this way you don't need a flag!)

ArthurZucker reviewed (commented):
nice work!

alyosha-swamy committed:
Address review feedback for AFMoE implementation

alyosha-swamy committed:
Add flex attention support to AFMoE model

ArthurZucker committed:
Merge branch 'main' into add_afmoe_model

alyosha-swamy committed:
Fix expert_bias routing in AFMoE

alyosha-swamy committed:
Remove test-results directory

ArthurZucker committed:
Merge branch 'main' into add_afmoe_model

ArthurZucker commented on src/transformers/models/afmoe/modular_afmoe.py:
```diff
@@ -0,0 +1,640 @@
+        if self.moe_enabled:
+            self.mlp = AfmoeMoE(config)
+        else:
+            self.mlp = AfmoeMLP(config)
```
again here

ArthurZucker commented on tests/models/afmoe/__init__.py:
```diff
@@ -0,0 +1,14 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
stuff should be imported, see `llama`'s init

alyosha-swamy committed:
Address PR review feedback for AFMoE model

ArthurZucker commented on src/transformers/models/afmoe/modular_afmoe.py:
```diff
@@ -0,0 +1,648 @@
+        return top_scores, selected_experts
+
+
+class AfmoeExperts(nn.ModuleList):
```
you could just inherti from Mixtral or Qwen2Moe it should be the same no?

alyosha-swamy replied:
The checkpoint weight structure is different in AFMoE

ArthurZucker replied:
We have an online weight converter but now worries :)

ArthurZucker reviewed (approved):
Thanks, we tend to try and remove code path as much as possible, if not done here we 'll do it post release !

alyosha-swamy committed:
fix(afmoe): ensure RMSNorm output dtype matches input dtype)

ArthurZucker committed:
properly return attn weights

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_42168). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

ArthurZucker committed:
fix most tests

ArthurZucker committed:
cleanup
Remove shared expert if else as defaults to 2
Remove `route_norm` as it default to `True`.

Make test smaller faster

ArthurZucker committed:
fix input embeds api

ArthurZucker committed:
update rope API, smaller test and should be good to go

ArthurZucker committed:
oups wront place to skip unittest

ArthurZucker committed:
quality

ArthurZucker committed:
update

ArthurZucker committed:
rope parameter docstring fill
